internal/internal_workflow_client.go

```diff
@@ -1708,45 +1765,104 @@ func (w *workflowClientInterceptor) UpdateWorkflow(
 			},
 		},
 	})
-	handle := &updateHandle{ref: resp.GetUpdateRef()}
 	if err != nil {
-		handle.err = err
-		return handle, nil
+		return nil, err
 	}
 	switch v := resp.GetOutcome().GetValue().(type) {
 	case nil:
-		panic("unspported update outcome: Incomplete")
+		return &lazyUpdateHandle{
+			client:           w.client,
+			baseUpdateHandle: baseUpdateHandle{ref: resp.GetUpdateRef()},
+		}, nil
 	case *updatepb.Outcome_Failure:
-		handle.err = w.client.failureConverter.FailureToError(v.Failure)
+		return &completedUpdateHandle{
+			err:              w.client.failureConverter.FailureToError(v.Failure),
+			baseUpdateHandle: baseUpdateHandle{ref: resp.GetUpdateRef()},
+		}, nil
 	case *updatepb.Outcome_Success:
-		handle.value = newEncodedValue(v.Success, w.client.dataConverter)
+		return &completedUpdateHandle{
+			value:            newEncodedValue(v.Success, w.client.dataConverter),
+			baseUpdateHandle: baseUpdateHandle{ref: resp.GetUpdateRef()},
+		}, nil
+	}
+	return nil, fmt.Errorf("unsupported outcome type %T", resp.GetOutcome().GetValue())
+}
+
+func (w *workflowClientInterceptor) PollWorkflowUpdate(
+	parentCtx context.Context,
+	in *ClientPollWorkflowUpdateInput,
+) (converter.EncodedValue, error) {
+	// header, _ = headerPropagated(ctx, w.client.contextPropagators)
+	//todo header not in PollWorkflowUpdate
+
+	pollReq := workflowservice.PollWorkflowExecutionUpdateRequest{
+		Namespace: w.client.namespace,
+		UpdateRef: in.UpdateRef,
+		Identity:  w.client.identity,
+		WaitPolicy: &updatepb.WaitPolicy{
+			LifecycleStage: enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_COMPLETED,
+		},
+	}
+	for parentCtx.Err() == nil {
+		ctx, cancel := newGRPCContext(
+			parentCtx,
+			grpcLongPoll(true),
+			grpcTimeout(pollUpdateTimeout),
+		)
+		ctx = context.WithValue(
+			ctx,
+			retry.ConfigKey,
+			createDynamicServiceRetryPolicy(ctx).GrpcRetryConfig(),
+		)
+		resp, err := w.client.workflowService.PollWorkflowExecutionUpdate(ctx, &pollReq)
+		cancel()
+		if err == context.DeadlineExceeded || status.Code(err) == codes.DeadlineExceeded {
```

Comment:
How do you know this isn't a legitimate deadline-exceeded from the server? Or how do you know this isn't a user-defined context deadline that we should bubble out? I think we need an empty response from the server here before SDK timeout. We don't need error codes appearing on normal operation. SDK side timeout IMO should only occur on actual error.

Reply:
I think I'm missing your intent here ... all DeadlineExceededs from the server are legitimate?

If it's the parent context that failed, we won't loop through again (see guard on for loop) - I added a test for the same.

Reply:
>all DeadlineExceededs from the server are legitimate?
Is there not way to tell the difference between the server not responding and the update not being complete yet?

Reply:
If you're fetching the result of an update, I don't think there exists a meaningful difference.

Reply:
My suggestion is very simply when an update is not done yet, do not return an error, return some other thing that says not done yet. Then let all timeout errors bubble as they may (whether user-defined or defined by SDK as longer than the poll-success-but-empty-try-again time).

Reply:
So an api like

```
Handle {
    Get(ctx context.Context, outPtr interface{})(error, bool)
}
```

?

I have to say that for me that's too weirdly different from Future. I understand that this api isn't a future but a somewhat semantically symmetric api is worth something.

Reply:
I am only talking about internal behavior here. Sorry for the confusion. Don't use gRPC errors for not-yet-keep-trying timeout from the server. The external user-facing Go API should remain the same. This comment is specifically on `if err == context.DeadlineExceeded || status.Code(err) == codes.DeadlineExceeded {` and I don't think we should always assume those mean retry.

Reply:
Okay so how about this: given that the server is currently implemented to return DEADLINE_EXCEEDED and that this was specifically chosen by that team, I don't want to remove the error check. That said, I can add a check for `outcome == nil` here and _also_ treat that as "not yet" so that if the server changes in the future we'll have that escape hatch. Thoughts on a compromise?

Reply:
> I don't want to remove the error check

So what if I call get update with my own context that has a deadline in 10s? You will not respect that. I think we need to revisit how it's currently implemented. I think it should be like all other poll calls IIUC in that it is not an error to not be done yet.

Reply:
> I think we need to revisit how it's currently implemented.

In general I agree with Chad that the server design in strange and should be looked at again, but I don't think this PR is the right forum to discuss a server design decision.

Reply:
> This code will not respect that and retry

The ctx used in the grpc call is derived from the caller supplied context so I'm not sure where you get that idea? Unless something in the internals is discarding the parentCtx. One hopes not. But I can do another unit test to confirm.

Reply:
I updated the unit test to confirm that if a caller submits a PollWorkflowUpdate with a context timeout of 50ms, that is indeed the timeout that is propagated through to the server and that if no user-supplied timeout is given, the default is used.

Reply:
I assume my `temporal_long_request_failure` metrics are gonna start going up if my update, say, takes 3 minutes?

Reply:
(a) can we run the this-code-will-not-respect-the-passed-context argument to ground before shifting? I really don't see in the code where that's the case and am asking you to make that clear

(b) re: the failure metric - I have no idea and will look into it. But again, the server's intention is to use a DEADLINE_EXCEEDED here so we may need to make allowances like we've done with longPoll.

Reply:
> can we run the this-code-will-not-respect-the-passed-context argument to ground before shifting? 

I think your context check in the loop is enough. So we can shift now. User context is no longer a concern. It just popped in my head as the first of many concerns about reusing errors for flow here.

> But again, the server's intention is to use a DEADLINE_EXCEEDED here so we may need to make allowances like we've done with longPoll.

Can we run the "server's intention" argument into the ground before shifting? Why is that the intention? Why can that not change? Do they do that on other long polls?

Reply:
It can absolutely change but you're talking about something that was discussed and settled a few weeks ago so it's not the sort of thing that's going to happen in the context of this PR.

Reply:
:+1: I vote for changing it. If it can't change now, ok, but I think it should. I think we'll be happier for it.

Reply:
If it helps, the code change to treat a nil response.Outcome as a signal to retry (subject to the parent context, natch) means that we can change the server without breaking field SDKs.